client/protocol.go

```diff
@@ -0,0 +1,413 @@
+func BuildCurpClientFromAddrs(addrs []string) (*protocolClient, error) {
+	clientTimeout := NewDefaultClientTimeout()
+	cluster, err := fetchCluster(addrs, clientTimeout.proposeTimeout)
+	if err != nil {
+		return nil, err
+	}
+
+	return &protocolClient{
+		addrs:    addrs,
+		connects: cluster.AllMembers,
+		timeout:  clientTimeout,
+	}, nil
+}
+
+// Fetch cluster from server
+func fetchCluster(addrs []string, proposeTimeout time.Duration) (*curpapi.FetchClusterResponse, error) {
+	var protocolClients []curpapi.ProtocolClient
+	fetchClusterRespCh := make(chan *curpapi.FetchClusterResponse)
+
+	for _, addr := range addrs {
+		conn, err := grpc.Dial(addr, grpc.WithTransportCredentials(insecure.NewCredentials()), grpc.WithIdleTimeout(1*time.Second))
+		if err != nil {
+			logrus.Warn(err)
+		} else {
+			protocolClient := curpapi.NewProtocolClient(conn)
+			protocolClients = append(protocolClients[:], protocolClient)
+		}
+	}
+
+	for _, protocolClient := range protocolClients {
+		client := protocolClient
+		go func() {
+			ctx, cancel := context.WithTimeout(context.Background(), proposeTimeout)
+			defer cancel()
+			res, err := client.FetchCluster(ctx, &curpapi.FetchClusterRequest{})
+			if err != nil {
+				logrus.Debug(err)
+			} else {
+				fetchClusterRespCh <- res
+			}
+		}()
+	}
+
+	select {
+	case fetchClusterRes := <-fetchClusterRespCh:
+		return fetchClusterRes, nil
+	case <-time.After(proposeTimeout):
+		return nil, errors.New("fetch cluster fail")
+	}
+}
+
+// Propose the request to servers, if use_fast_path is false, it will wait for the synced index
+func (c protocolClient) Propose(cmd *xlineapi.Command, useFastPath bool) (*ProposeResponse, error) {
+	fastResCh := make(chan *ProposeResponse)
+	fastErrCh := make(chan error)
+	slowResCh := make(chan *ProposeResponse)
+	slowErrCh := make(chan error)
+
+	go func() {
+		proposeRes, err := c.slowRound(cmd)
+		slowResCh <- proposeRes
+		slowErrCh <- err
+	}()
+
+	go func() {
+		proposeRes, err := c.fastRound(cmd)
+		fastResCh <- proposeRes
+		fastErrCh <- err
+	}()
+
+	if useFastPath {
+		select {
+		case proposeRes := <-fastResCh:
+			err := <-fastErrCh
+			if proposeRes.IsFastRoundSuccess {
+				return proposeRes, err
+			}
+			proposeRes = <-slowResCh
+			err = <-slowErrCh
+			return proposeRes, err
+		case proposeRes := <-slowResCh:
+			err := <-slowErrCh
+			return proposeRes, err
+		}
+	} else {
+		<-fastResCh
+		<-fastErrCh
+		proposeRes := <-slowResCh
+		err := <-slowErrCh
+		return proposeRes, err
+	}
+}
+
+// The fast round of Curp protocol
+// It broadcast the requests to all the curp servers.
+func (c protocolClient) fastRound(cmd *xlineapi.Command) (*ProposeResponse, error) {
+	logrus.Info("Fast round start. Propose ID:", cmd.ProposeId)
+
+	isReceivedLeaderReq := false
+	isReceivedSuccessRes := 0
+	okCnt := 0
+	var clients []curpapi.ProtocolClient
+	var cmdRes xlineapi.CommandResponse
+	var proposeErr xlineapi.ExecuteError
```

Comment:
Please don't name it `proposeErr` here, it will cause confusion.

Reply:
done